pkg/agent/packetcapture/packetcapture_controller_test.go

```diff
@@ -0,0 +1,569 @@
+			"username": []byte("AAA"),
+			"password": []byte("BBBCCC"),
+		},
+	}
+}
+
+func genTestCR(name string, num int32) *crdv1alpha1.PacketCapture {
+	result := &crdv1alpha1.PacketCapture{
+		ObjectMeta: metav1.ObjectMeta{Name: name, UID: types.UID(fmt.Sprintf("uid-%s", name))},
+		Spec: crdv1alpha1.PacketCaptureSpec{
+			Source: crdv1alpha1.Source{
+				Pod: &crdv1alpha1.PodReference{
+					Namespace: pod1.Namespace,
+					Name:      pod1.Name,
+				},
+			},
+			Destination: crdv1alpha1.Destination{
+				Pod: &crdv1alpha1.PodReference{
+					Namespace: pod2.Namespace,
+					Name:      pod2.Name,
+				},
+			},
+			CaptureConfig: crdv1alpha1.CaptureConfig{
+				FirstN: &crdv1alpha1.PacketCaptureFirstNConfig{
+					Number: num,
+				},
+			},
+			Packet: &crdv1alpha1.Packet{
+				Protocol: &icmpProto,
+			},
+			FileServer: &crdv1alpha1.PacketCaptureFileServer{
+				URL: "sftp://127.0.0.1:22/aaa",
+			},
+		},
+	}
+	return result
+}
+
+type testUploader struct {
+	url      string
+	fileName string
+}
+
+func (uploader *testUploader) Upload(url string, fileName string, config *ssh.ClientConfig, outputFile io.Reader) error {
+	if url != uploader.url {
+		return fmt.Errorf("expected url: %s for uploader, got: %s", uploader.url, url)
+	}
+	if fileName != uploader.fileName {
+		return fmt.Errorf("expected filename: %s for uploader, got: %s", uploader.fileName, fileName)
+	}
+	return nil
+}
+
+func craftTestPacket() gopacket.Packet {
+	buffer := gopacket.NewSerializeBuffer()
+	options := gopacket.SerializeOptions{}
+	rawBytes := []byte{10, 20, 30}
+	gopacket.SerializeLayers(buffer, options,
+		&layers.Ethernet{
+			SrcMAC: net.HardwareAddr{0xFF, 0xAA, 0xFA, 0xAA, 0xFF, 0xAA},
+			DstMAC: net.HardwareAddr{0xBD, 0xBD, 0xBD, 0xBD, 0xBD, 0xBD},
+		},
+		&layers.IPv4{
+			SrcIP: net.IP{127, 0, 0, 1},
+			DstIP: net.IP{8, 8, 8, 8},
+		},
+		&layers.TCP{
+			SrcPort: layers.TCPPort(4321),
+			DstPort: layers.TCPPort(80),
+		},
+		gopacket.Payload(rawBytes),
+	)
+	return gopacket.NewPacket(buffer.Bytes(), layers.LayerTypeEthernet, gopacket.NoCopy)
+}
+
+type testCapture struct {
+}
+
+func (p *testCapture) Capture(ctx context.Context, device string, srcIP, dstIP net.IP, packet *crdv1alpha1.Packet) (chan gopacket.Packet, error) {
+	ch := make(chan gopacket.Packet, 15)
+	for i := 0; i < 15; i++ {
+		ch <- craftTestPacket()
+	}
+	return ch, nil
+}
+
+type fakePacketCaptureController struct {
+	*Controller
+	kubeClient         kubernetes.Interface
+	mockController     *gomock.Controller
+	crdClient          *fakeversioned.Clientset
+	crdInformerFactory crdinformers.SharedInformerFactory
+	informerFactory    informers.SharedInformerFactory
+}
+
+func newFakePacketCaptureController(t *testing.T, runtimeObjects []runtime.Object, initObjects []runtime.Object) *fakePacketCaptureController {
+	controller := gomock.NewController(t)
+	objs := []runtime.Object{
+		&pod1,
+		&pod2,
+		&pod3,
+		&secret1,
+	}
+	objs = append(objs, generateTestSecret())
```

Comment:
I don't understand why we need `generateTestSecret()` when we already have `secret1`?